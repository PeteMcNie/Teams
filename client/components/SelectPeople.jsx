import React from 'react'
import { connect } from 'react-redux'

import Loading from './Loading'

import { shuffleSelectedPeople } from '../sortFunctions'
import { getPeopleAction, deletePerson, teamCreator } from '../actions'

class SelectPeople extends React.Component {
_isMounted = false

state = {
  selectedPeople: [],
  numberOfTeams: '2'
}

componentDidMount () {
  this._isMounted = true
  this.props.dispatch(getPeopleAction())
    .then(() => {
      if (this._isMounted) {
        this.setState({
          people: this.props.people
        })
      }
    })
}

componentWillUnmount () {
  this._isMounted = false
}

    handleCheck = (evnt, person) => {
      const isChecked = evnt.target.checked
      let selected = [...this.state.selectedPeople]
      if (isChecked) {
        selected.push(person)
      } else {
        selected = selected.filter(i => person !== i)
      }
      this.setState({
        selectedPeople: selected
      })
    }

    handleCheckAll = () => {
      if (this.state.selectedPeople.length === this.state.people.length) {
        this.setState({
          selectedPeople: []
        })
      } else {
        this.setState({
          selectedPeople: this.state.people
        })
      }
    }

    handleSelect = evt => {
      this.setState({
        numberOfTeams: evt.target.value
      })
    }

    deletePerson = (evt, id) => {
      evt.preventDefault()
      this.props.dispatch(deletePerson(id))
      this.props.dispatch(getPeopleAction())
        .then(() => {
          this.setState({
            people: this.props.people
          })
        })
    }

    handleSubmit = evnt => {
      evnt.preventDefault()

      const selected = this.state.selectedPeople
      const numberOfTeams = this.state.numberOfTeams
      // console.log('Selected people in selectedPeople.jsx: ', selected)

      const teams = shuffleSelectedPeople(selected, numberOfTeams)
      // console.log('selectPeople ', teams)

      this.props.dispatch(teamCreator(teams))
      this.props.history.push('/teams')
    }

    render () {
      console.log('FOR TEST ', this.props.people)
      if (this.props.loading) {
        return <Loading />
      } else {
        return (
          <div>
            <form onSubmit={this.handleSubmit}>
              <ul>
                {this.props.people.map(person => {
                  return (
                    <li key={person.id}>
                      <input
                        type='checkbox'
                        name={person.name}
                        checked={this.state.selectedPeople.includes(person)}
                        onChange={(evnt) => this.handleCheck(evnt, person)}
                      />  {person.name}
                      <button onClick={evt => this.deletePerson(evt, person.id)}><i className="fas fa-trash"></i></button>
                    </li>
                  )
                }
                )}
              </ul>
              <div>
                <input type='button' onClick={this.handleCheckAll} value='Select / Deselect All' />
              </div>
              <div>
                <p>Select number of Teams</p>
                <select name='amount' value={this.state.numberOfTeams} onChange={this.handleSelect}>
                  {[2, 3, 4, 5, 6, 7, 8, 9, 10].map(i => <option key={i} value={i}>{i}</option>)}
                </select>
              </div>
              <div>
                <input type='submit' value='Create Teams!' />
              </div>
            </form>
          </div>
        )
      }
    }
}

function mapStateToProps (state) {
  return {
    loading: state.loading,
    people: state.getPeopleReducer
  }
}

export default connect(mapStateToProps)(SelectPeople)
