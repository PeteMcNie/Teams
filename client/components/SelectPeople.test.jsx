import React from 'react'

import { renderWithRedux } from '../testing/utils'
import { screen } from '@testing-library/react'

import '@testing-library/jest-dom'

import SelectPeople from './SelectPeople'

const getPeopleReducer = [
  { id: 1, name: 'Pete', isSelected: 0 },
  { id: 2, name: 'Nigel', isSelected: 0 },
  { id: 3, name: 'Andre', isSelected: 0 },
  { id: 4, name: 'Sergio', isSelected: 0 }
]

jest.mock('../actions', () => {
  return {
    getPeopleAction: () => () => Promise.resolve()
  }
})

test('SelectPeople component matchs snapshot', () => {
  const { container } = renderWithRedux(<SelectPeople />)
  expect(container).toMatchSnapshot()
})

test('Select People component renders users', async () => {
  renderWithRedux(<SelectPeople />, { initialState: { getPeopleReducer } })

  const firstPerson = await screen.getByText('Pete')
  expect(firstPerson).toBeInTheDocument()

  const secondPerson = await screen.getByText('Nigel')
  expect(secondPerson).toBeInTheDocument()

  const forthPerson = await screen.getByText('Sergio')
  expect(forthPerson).toBeInTheDocument()
})

test('Select People component renders buttons correctly', async () => {
  renderWithRedux(<SelectPeople />)

  expect(screen.getByRole('button', { name: 'Select / Deselect All' })).toBeInTheDocument()
  expect(screen.getByRole('button', { name: 'Submit' })).toBeInTheDocument()
})

test('Select People component renders the correct options', async () => {
  renderWithRedux(<SelectPeople />)

  expect(screen.getByRole('option', { name: '6' })).toBeInTheDocument()
  expect(screen.queryByText('11')).not.toBeInTheDocument()
})
