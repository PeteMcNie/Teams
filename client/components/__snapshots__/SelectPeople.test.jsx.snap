// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectPeople component matchs snapshot 1`] = `
<div>
  <div>
    <form>
      <ul
        class="peopleGrid"
        data-cy="peopleList"
      />
      <div>
        <input
          class="selectAll"
          type="button"
          value="Select / Deselect All"
        />
      </div>
      <div>
        <p
          class="heading"
        >
          Select number of Teams
        </p>
        <select
          class="numberOfTeams"
          data-cy="teamNumber"
          name="amount"
        >
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </div>
      <div>
        <input
          class="createTeams"
          data-cy="submit"
          type="submit"
          value="Create Teams!"
        />
      </div>
    </form>
  </div>
</div>
`;
